### ISSUE TITLE
Could not parse object/Your branch and 'origin/master' have diverged

### ISSUE THREAD

**Issue by skapral** [yegor256/0pdd#99]
https://github.com/project-avral/oo-atom/commit/9c5f8238a9ef58a280fced6283858886f27166f5#commitcomment-22772561

> fatal: Could not parse object '251000d38852773d78985b00a29c6685808bd0ca^'
> Unknown exit code (128) from command: git-merge-recursive 251000d38852773d78985b00a29c6685808bd0ca^ -- HEAD 251000d38852773d78985b00a29c6685808bd0ca
> 
> Your branch and 'origin/master' have diverged,
> and have 1 and 3 different commits each, respectively.
>   (use "git pull" to merge the remote branch into yours)
> First, rewinding head to replay your work on top of it...

I suspect that root cause of this issue is that I did some force pushes earlier. I aware that it is a sin, but the main question is: what can I do now to fix this issue?

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this issue

**Comment by yegor256** [comment]
@skapral I'm aware of this problem and will try to fix ASAP. I don't know how, though :) #98 is relevant

**Comment by skapral** [comment]
@yegor256 thanks for feedback. About the root cause - did I guessed right? Are there any known workarounds?

**Comment by yegor256** [comment]
@skapral yes, the root cause is your "forced push". The problem is that I don't know how to reproduce this trouble in the test.

**Comment by skapral** [comment]
I'll try. However, I'm not a Ruby guy, so it may take a while.

**Comment by skapral** [comment]
Ok. I give up. It is not reproduced for me either. I gave another try to my project and the issue is not reproduced.

Probably we need a button on the project's page on 0pdd.com, like "drop corrupted repo from 0pdd and start anew"? What do you think?

**Comment by yegor256** [comment]
@0crat assign @tedtoer

**Comment by 0crat** [comment]
> @0crat assign @tedtoer  ([here](https://github.com/yegor256/0pdd/issues/99#issuecomment-336805431))

@yegor256 Job `gh:yegor256/0pdd#99` assigned to @tedtoer, please go ahead ([policy](http://datum.zerocracy.com/pages/policy.html)).

**Comment by 0crat** [comment]
@tedtoer this job was assigned to you 8 days ago.
It will be taken away from you after 10 days from start (this is our policy).

**Comment by 0crat** [comment]
@tedtoer resigned from `gh:yegor256/0pdd#99`, please stop working.

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/0pdd#99` is not assigned to anyone.

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/0pdd#99` is not assigned to anyone.

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/0pdd#99` is not assigned to anyone.

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/0pdd#99` is not assigned to anyone.

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/0pdd#99` is not assigned to anyone.

**Comment by 0crat** [comment]
@tedtoer resigned from `gh:yegor256/0pdd#99`, please stop working.

**Comment by 0crat** [comment]
Job `gh:yegor256/0pdd#99` assigned to @tedtoer. The budget is [fixed](http://datum.zerocracy.com/pages/policy.html#4) and it is 30 minutes. Please, read the [Policy](http://datum.zerocracy.com/pages/policy.html) and go ahead.

**Comment by 0crat** [comment]
@tedtoer resigned from `gh:yegor256/0pdd#99`, please stop working.

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/0pdd/issues/99#issuecomment-360779327))

@yegor256 This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#99](https://github.com/yegor256/0pdd/issues/99) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 3mon
  * The role is `DEV`
  * The job is not assigned to anyone
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 30 minutes
  * These users are banned and won't be assigned:
    * @skapral[/z](https://www.0crat.com/u/skapral): This user reported the ticket
    * @tedtoer[/z](https://www.0crat.com/u/tedtoer): User was resigned from the ticket
    * @tedtoer[/z](https://www.0crat.com/u/tedtoer): User was resigned from the ticket

**Comment by 0crat** [comment]
The job [#99](https://github.com/yegor256/0pdd/issues/99) assigned to @valentjedi[/z](https://www.0crat.com/u/valentjedi), here is [why](https://www.0crat.com/footprint/C3T46CUJJ/441843943). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by 0crat** [comment]
The user @valentjedi[/z](https://www.0crat.com/u/valentjedi) resigned from [#99](https://github.com/yegor256/0pdd/issues/99), please stop working

**Comment by 0crat** [comment]
Resigned on delay, see [§8](http://www.zerocracy.com/policy.html#8): -30 points just awarded to @valentjedi[/z](https://www.0crat.com/u/valentjedi), total is [-120](https://www.0crat.com/u/valentjedi)

**Comment by 0crat** [comment]
The job [#99](https://github.com/yegor256/0pdd/issues/99) assigned to @prondzyn[/z](https://www.0crat.com/u/prondzyn), here is [why](https://www.0crat.com/footprint/C3T46CUJJ/641245441). The budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by yegor256** [comment]
@0crat out

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/yegor256/0pdd/issues/99#issuecomment-370479878))

@yegor256 The job [#99](https://github.com/yegor256/0pdd/issues/99) is now out of scope

**Comment by skapral** [comment]
@yegor256, I am ok with closing it, I never saw it again later.

But what about my alternative proposal:

> Probably we need a button on the project's page on 0pdd.com, like "drop corrupted repo from 0pdd and start anew"? What do you think?

Despite the fact that the issue is not reproducible anymore, it is obvious from logs that the fault came from locally cloned Git repo. This button would allow the user to unblock himself this or similar issues.

What do you think - is it worth to outline this idea to a separate bug?

**Comment by yegor256** [comment]
@skapral but the problem was solved. Now, no matter how corrupted is the local repo, the next push will restore it back to the normal/current state. It has to happen automatically. If it doesn't happen, it's a bug. To my knowledge, we don't have such bugs now. Maybe they will show up in the future.

**Comment by skapral** [comment]
@yegor256 Initially, when I observed this bug first time, I clearly remember that this "next push" didn't worked for me. IMO if the local 0pdd clone became corrupted, the most reliable strategy to recover from it is to make 0pdd re-clone the local repo anew, that's why I proposed that button.

But ok, since the issue is not reproduced anymore, let's close it.

**Comment by yegor256** [comment]
@skapral a fresh `clone` is not the only option. Now we are doing this and it seem to work:

https://github.com/yegor256/0pdd/blob/0.29.3/objects/git_repo.rb#L99-L108
